bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+            id,
+            config,
+            peers: BTreeSet::new(),
+            clock: 0,
+            rack_secret_state: RackSecretState::Empty,
+        }
+    }
+
+    // Round-robin peer selection
+    pub fn next_peer(&self, current: &Baseboard) -> Option<Baseboard> {
+        if let Some(index) = self.peers.iter().position(|x| x == current) {
+            let next_index = (index + 1) % self.peers.len();
+            self.peers.iter().nth(next_index).cloned()
+        } else {
+            self.peers.first().cloned()
+        }
+    }
+
+    // Send a `GetShare` request to the newly connected peer if necessary
+    pub fn on_connect(&self, peer: Baseboard, rack_uuid: Uuid) -> Output {
+        if let RackSecretState::Retrieving { request_id, from, .. } =
+            &self.rack_secret_state
+        {
+            // We don't have a share from the peer and we weren't previously
+            // connected to the peer.
+            if !from.contains(&peer) && !self.peers.contains(&peer) {
+                let request = Request {
+                    id: *request_id,
+                    type_: RequestType::GetShare { rack_uuid },
+                };
+                return Output {
+                    persist: false,
+                    envelopes: vec![Envelope { to: peer, msg: request.into() }],
+                    api_output: None,
+                };
+            }
+        }
+        Output::none()
+    }
+}
+
+/// Metadata associated with a request that is keyed by `Baseboard`
+#[derive(Debug)]
+pub struct RequestMetadata {
+    pub request_id: Uuid,
+    pub expiry: Ticks,
+}
+
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct Shares(Vec<Vec<u8>>);
+
+impl Shares {
+    fn insert(&mut self, share: Vec<u8>) {
+        self.0.push(share);
+    }
+}
+
+impl Debug for Shares {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Shares").finish()
+    }
+}
+
+/// The state of rack secret share retrieval
+///
+/// Primarily this is whether we are in the process of retrieving enough shares to
+/// recompute the rack secret or if we have already recomputed it.
+#[derive(Debug)]
+pub enum RackSecretState {
+    /// We are not currently retrieving any shares and we don't have a computed
+    /// rack secret
+    Empty,
+
+    /// We only store the shares in memory until some expiry deadline.
+    /// However, we extend this deadline each time a share is learned to enable
+    /// computation of the rack secret. If no shares are added for a while, and
+    /// we still don't have enough shares to compute the rack secret then we
+    /// drop all the shares.
+    Retrieving {
+        // We use a separate BTreeSet to track who we retrieved shares form so
+        // that we can wrap the shares with a `Secret`
+        request_id: Uuid,
+        from: BTreeSet<Baseboard>,
+        shares: Shares,
+        expiry: Ticks,
+        threshold: usize,
+        share_digests: Vec<Sha3_256Digest>,
```

Comment:
Does this always have to be the same order / length as `shares`?  If so, maybe store them together?

Reply:
Ah no, this is used to validate incoming shares.  In that case, it should be a `BTreeSet` for efficient searching, since we call `share_digests.contains`